core/iwasm/aot/aot_loader.c

```diff
@@ -2532,6 +2553,76 @@ load_init_data_section(const uint8 *buf, const uint8 *buf_end,
     return false;
 }
 
+static bool
+merge_data_and_text(const uint8 **buf, const uint8 **buf_end, AOTModule *module,
+                    char *error_buf, uint32 error_buf_size)
+{
+    uint8 *old_buf = (uint8 *)*buf;
+    uint8 *old_end = (uint8 *)*buf_end;
+    size_t code_size = (size_t)(old_end - old_buf);
+    uint32 page_size = os_getpagesize();
+    uint64 total_size = 0;
+    uint32 i;
+    uint8 *sections;
+    if (code_size == 0) {
+        return true;
+    }
+    /* calc total memory needed */
+    total_size += ((uint64)code_size + page_size - 1) & ~(page_size - 1);
+    for (i = 0; i < module->data_section_count; ++i) {
+        total_size += ((uint64)module->data_sections[i].size + page_size - 1)
+                      & ~(page_size - 1);
+    }
+    if (total_size > (uint64)code_size) {
```

Comment:
Should be `total_size > 0`? os_mmap should be called no matter that there is data section or not.